nuxeo-services/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/requestcontroller/filter/NuxeoCorsCsrfFilter.java

```diff
@@ -0,0 +1,191 @@
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpHead;
+import org.apache.http.client.methods.HttpOptions;
+import org.apache.http.client.methods.HttpTrace;
+import org.nuxeo.ecm.platform.web.common.requestcontroller.service.RequestControllerManager;
+import org.nuxeo.ecm.platform.web.common.vh.VirtualHostHelper;
+import org.nuxeo.runtime.api.Framework;
+
+import com.thetransactioncompany.cors.CORSConfiguration;
+import com.thetransactioncompany.cors.CORSFilter;
+import com.thetransactioncompany.cors.Origin;
+
+/**
+ * Nuxeo CORS and CSRF filter, returning CORS configuration and preventing CSRF attacks by rejecting dubious requests.
+ *
+ * @since 5.7.2 for CORS
+ * @since 10.1 for CSRF
+ */
+public class NuxeoCorsCsrfFilter implements Filter {
+
+    private static final Log log = LogFactory.getLog(NuxeoCorsCsrfFilter.class);
+
+    public static final String GET = HttpGet.METHOD_NAME;
+
+    public static final String HEAD = HttpHead.METHOD_NAME;
+
+    public static final String OPTIONS = HttpOptions.METHOD_NAME;
+
+    public static final String TRACE = HttpTrace.METHOD_NAME;
+
+    @Override
+    public void init(FilterConfig filterConfig) {
+        // nothing to do
+    }
+
+    @Override
+    public void destroy() {
+        // nothing to do
+    }
+
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain chain)
+            throws IOException, ServletException {
+        HttpServletRequest request = (HttpServletRequest) servletRequest;
+        HttpServletResponse response = (HttpServletResponse) servletResponse;
+
+        RequestControllerManager service = Framework.getService(RequestControllerManager.class);
+        CORSFilter corsFilter = service.getCorsFilterForRequest(request);
+        CORSConfiguration corsConfig = corsFilter == null ? null : corsFilter.getConfiguration();
+        String method = request.getMethod();
+        URI sourceURI = getSourceURI(request);
+        URI targetURI = getTargetURI(request);
+        if (log.isDebugEnabled()) {
+            log.debug("Method: " + method + ", source: " + sourceURI + ", target: " + targetURI);
+        }
+
+        boolean allow;
+        if (GET.equals(method) || HEAD.equals(method) || OPTIONS.equals(method) || TRACE.equals(method)) {
+            // safe method according to RFC 7231 4.2.1
+            log.debug("Safe method: allow");
+            allow = true;
+        } else if (sourceAndTargetMatch(sourceURI, targetURI)) {
+            // source and target match, or not provided
+            log.debug("Source and target match: allow");
+            if (targetURI == null) {
+                // misconfigured server or proxy headers
+                log.error("Cannot determine target URL for CSRF check");
+            }
+            allow = true;
+        } else if (corsFilter == null) {
```

Comment:
Replace or add `corsConfig == null` ?

Reply:
Ah yes this is nicer. This code was refactored so many times during dev that I missed this nice cleanup.